docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md

```diff
@@ -0,0 +1,245 @@
+# Build-time and runtime detection for hardware acceleration with advance CPU features
+
+This document describes how to support the algorithms that can do hardware acceleration with advance CPU features.
+
+## Introduction
+
+As description in [Build on Arm64 HOST Fail](https://github.com/Mbed-TLS/mbedtls/issues/5758), we got two issues that can be fixed with [#6932](https://github.com/Mbed-TLS/mbedtls/pull/6932)
+- `check_config.h` reports fail when HOST-CC is Arm64
+- With `-march=armv8.2-a+sha3` cpu flags
+Until [#6895](https://github.com/Mbed-TLS/mbedtls/pull/6895), we forbiden enable `MBEDTLS_SHA512_USE_A64_IF_PRESENT` and `MBEDTLS_AESCE_C` to workaround it.
+
+For time being, we have 4 hardware accelerations in `development` and 1 in [#6918](https://github.com/Mbed-TLS/mbedtls/pull/6918)
+- SHA256 with Arm64 crypto extension
+- SHA512 with Arm64 sha3 extension
+- AES/GCM with x86_64 aes extension
+- AES/GCM with x86 padlock extension
+- AES/GCM with Arm64 crypto extension
+There are 5 functions for `runtime detection`. We can remove two functions. For Arm64, the detection function in SHA512 is the best one.
+And SHA{512,256} implement Arm64 hardware acceleration only. AES/GCM does not implmented it
+
+As my understand, we should meet below requirements.
+- With `runtime detection`.
+  - `runtime detection` for architecture MUST be enabled.
+  - Enable/disable hardware accelerations with configure options.
+  - Enable/disable `no hardware accelerations` with configure options.
+- Without `runtime detection`( Build-time detection only).
+  - `runtime detection` for architecture MUST be disabled.
+  - If `no hardware accelerations` is enabled, all hardwares accelerations MUST be disabled.
+  - If `no hardware accelerations` is disabled, only one hardwares acceleration can be enabled.
+  - If possible, enable/disable hardware accelerations base on `CFLAGS`.
+
+NOTE: The algorithm can be implemented with different instruction sets on same architecture. For example, AES can be implemented with crypto extention or SVEAES extension in Arm64.
+
+
+## Suggestion
+
+### check config
+
+`check_config.h` MUST not include any archeticture special checks. This gives up the advantage of the current code. But that makes it architecture independent.
```

Comment:
>On a related note, `check_config.h` must not check compiler options.

Do you mean it has been included in other document? I did not get it when I wrote it.

Reply:
`check compiler options` can reduce the risk of runtime failure. But I do not think it is a big problem.

Reply:
I don't think we've ever written a formal document to say “`check_config.h` must not check compiler options”, because that wasn't the purpose of `check_config.h`. Its intent is to check preprocessor definitions.

Reply:
it is removed

Reply:
I moved some cflags checks to source file in #6932 . Please take a look